geomesa-z3/src/main/scala/org/locationtech/geomesa/curve/Z3.scala

```diff
@@ -107,52 +108,66 @@ object Z3 {
    * Recurse down the oct-tree and report all z-ranges which are contained
    * in the cube defined by the min and max points
    */
-  def zranges(min: Z3, max: Z3, maxRecurse: Int): Seq[(Long, Long)] = {
-    var mq: MergeQueue = new MergeQueue
-    val sr = Z3Range(min, max)
-
-    var recCounter = 0
-    var reportCounter = 0
-
-    def zranges(prefix: Long, offset: Int, quad: Long, level: Int): Unit = {
-      recCounter += 1
-
-      val min: Long = prefix | (quad << offset) // QR + 000..
-      val max: Long = min | (1L << offset) - 1 // QR + 111..
-      val qr = Z3Range(new Z3(min), new Z3(max))
-      if (level <= maxRecurse) {
-        if (sr containsInUserSpace qr) {
-          // whole range matches, happy day
-          mq += (qr.min.z, qr.max.z)
-          reportCounter += 1
-        } else if (sr overlapsInUserSpace qr) { // TODO move this?
-          if (offset > 0) {
-            // some portion of this range are excluded
-            zranges(min, offset - MAX_DIM, 0, level + 1)
-            zranges(min, offset - MAX_DIM, 1, level + 1)
-            zranges(min, offset - MAX_DIM, 2, level + 1)
-            zranges(min, offset - MAX_DIM, 3, level + 1)
-            zranges(min, offset - MAX_DIM, 4, level + 1)
-            zranges(min, offset - MAX_DIM, 5, level + 1)
-            zranges(min, offset - MAX_DIM, 6, level + 1)
-            zranges(min, offset - MAX_DIM, 7, level + 1)
-            //let our children punt on each subrange
-          } else {
-            mq += (qr.min.z, qr.max.z)
-          }
+  def zranges(min: Z3, max: Z3): Seq[(Long, Long)] = {
+    val (commonPrefix, commonBits) = longestCommonPrefix(min.z, max.z)
+
+    // base our recursion on the depth of the tree that we get 'for free' from the common prefix
+    val maxRecurse = if (commonBits < 30) 7 else if (commonBits < 40) 6 else 5
```

Comment:
It will get stopped either via the contains clause or via the offset > 0 clause.